Resolve 500 error with dissemination search/advanced

rnovak338 opened the pull request:
Addresses #4437.
- Add form error labels for start and end dates in search.
- Re-render search template instead of raising validation errors.

## How to test

1. Boot up the application.
2. Attempt to search with invalid date formats under "FAC Accepted Date" from the search and advanced search pages.
3. Each of the pages should reload with an error message over the date field.

## PR Checklist: Submitter

- [ ]   Link to an issue if possible. If there’s no issue, describe what your branch does. Even if there is an issue, a brief description in the PR is still useful.
- [ ]   List any special steps reviewers have to follow to test the PR. For example, adding a local environment variable, creating a local test file, etc.
- [ ]   For extra credit, submit a screen recording like [this one](https://github.com/GSA-TTS/FAC/pull/1821).
- [ ]   Make sure you’ve merged `main` into your branch shortly before creating the PR. (You should also be merging `main` into your branch regularly during development.)
- [ ]   Make sure you’ve accounted for any migrations. When you’re about to create the PR, bring up the application locally and then run `git status | grep migrations`. If there are any results, you probably need to add them to the branch for the PR. Your PR should have only **one** new migration file for each of the component apps, except in rare circumstances; you may need to delete some and re-run `python manage.py makemigrations` to reduce the number to one. (Also, unless in exceptional circumstances, your PR should not delete any migration files.)
- [ ]   Make sure that whatever feature you’re adding has tests that cover the feature. This includes test coverage to make sure that the previous workflow still works, if applicable.
- [ ]   Make sure the full-submission.cy.js [Cypress test](https://github.com/GSA-TTS/FAC/blob/main/docs/testing.md#end-to-end-testing) passes, if applicable.
- [ ]   Do manual testing locally. Our tests are not good enough yet to allow us to skip this step. If that’s not applicable for some reason, check this box.
- [ ]   Verify that no Git surgery was necessary, or, if it was necessary at any point, repeat the testing after it’s finished.
- [ ]   Once a PR is merged, keep an eye on it until it’s deployed to dev, and do enough testing on dev to verify that it deployed successfully, the feature works as expected, and the happy path for the broad feature area (such as submission) still works.
- [ ]   Ensure that prior to merging, the working branch is up to date with main and the terraform plan is what you expect.

## PR Checklist: Reviewer

- [ ]   Pull the branch to your local environment and run `make docker-clean; make docker-first-run && docker compose up`; then run `docker compose exec web /bin/bash -c "python manage.py test"`
- [ ]   Manually test out the changes locally, or check this box to verify that it wasn’t applicable in this case.
- [ ]   Check that the PR has appropriate tests. Look out for changes in HTML/JS/JSON Schema logic that may need to be captured in Python tests even though the logic isn’t in Python.
- [ ]   Verify that no Git surgery is necessary at any point (such as during a merge party), or, if it was, repeat the testing after it’s finished.

The larger the PR, the stricter we should be about these points.

## Pre Merge Checklist: Merger

- [ ]   Ensure that prior to approving, the terraform plan is what we expect it to be. `-/+ resource "null_resource" "cors_header" ` should be destroying and recreating its self and ` ~ resource "cloudfoundry_app" "clamav_api" ` might be updating its `sha256` for the `fac-file-scanner` and `fac-av-${ENV}` by default.
- [ ]   Ensure that the branch is up to date with `main`.
- [ ]   Ensure that a terraform plan has been recently generated for the pull request.

phildominguez-gsa commented:
This does work well, just a couple questions/suggestions!

No luck getting this to show up?
<img width="226" alt="image" src="https://github.com/user-attachments/assets/4700dc49-ceb9-4949-ba38-78fafb061ff9">

If it's not doable, I think we'd need to better indicate that the search failed due to an error. Right now you can't really tell until you expand the date section again. Maybe put a message where the search results show up, something like "Search submission failed" with a list of all the errors?

Ideally the form should retain all the info the user submitted as well, if possible. I could see it being pretty frustrating if everything got wiped because you entered a bad date 😬

rnovak338 commented:
> This does work well, just a couple questions/suggestions!
> 
> No luck getting this to show up, like in the step 2 form? <img alt="image" width="226" src="https://private-user-images.githubusercontent.com/142051477/391985866-4700dc49-ceb9-4949-ba38-78fafb061ff9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMzNTIzNDIsIm5iZiI6MTczMzM1MjA0MiwicGF0aCI6Ii8xNDIwNTE0NzcvMzkxOTg1ODY2LTQ3MDBkYzQ5LWNlYjktNDk0OS1iYTM4LTc4ZmFmYjA2MWZmOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwNFQyMjQwNDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MGFjZTFjYmExZmU3YTIyOTYwNzlkY2UyOWM2YzJjYzMyMTIzYWMwYmU0Yjk3NDdiYzgyMDEyMTFlODgzNTgyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.vLne2SY9ZAhqifZp1sxKTK8HOdJnmEpXtp0tx3Qx5No">
> 
> If it's not doable, I think we'd need to better indicate that the search failed due to an error. Right now you can't really tell until you expand the date section again. Maybe put a message where the search results show up, something like "Search submission failed" with a list of all the errors?
> 
> Ideally the form should retain all the info the user submitted as well, if possible. I could see it being pretty frustrating if everything got wiped because you entered a bad date 😬

Hey Phil,

Thanks for your feedback. That tooltip for the form field is generated by the browser, and it does not seem like we can easily guarantee a way to get it to generate without changing up some of the front-end logic quite a bit. I updated the PR with the following changes for both search AND advanced search. Please double check this on your end when you are ready!

New USWDS error component renders on invalid form field. It looks like this:
<img width="548" alt="image" src="https://github.com/user-attachments/assets/f7084b26-b0d5-4f08-9dc4-2560f2e248e5">

Your inputted fields will keep the values they had before submission (and the appropriate expanders are open/closed). The error will appear like this:
<img width="305" alt="image" src="https://github.com/user-attachments/assets/6ce61ff8-01c8-4454-ac7e-b37c9d54baa0">

Give it a test on both basic and advanced search.

jperson1 commented:
^ Relatedly, my browser won't give me an error the first time I put in a bad date. But if I get the bad date error from submitting and _then_ go in with a bad date, my browser will let me know. Not the first time, when I would have needed it. Weird. Oh well.

![image](https://github.com/user-attachments/assets/00fd3237-d2cd-40e9-8475-c0ea0947728e)

jperson1 commented on backend/dissemination/templates/search-alert-error.html:
```diff
@@ -0,0 +1,8 @@
+<div class="usa-alert usa-alert--error">
+    <div class="usa-alert__body">
+        <h2 class="usa-alert__heading">Error</h2>
+        <p>Oops! It looks like some of the information you entered is invalid. Please double-check your information for errors.</p>
```
Not sure about "Oops." Might be too friendly for the FAC's voice? @James-Paul-Mason, how do you feel?

We have a similar sort of message here:
https://github.com/GSA-TTS/FAC/blob/53c8abdb5e8c387d7de5ec987d430d96a1b8339f/backend/audit/templates/audit/manage-submission-change-access.html#L34-L36

phildominguez-gsa replied:
Once we get guidance from James on this I'm good with approving. Works great!

James-Paul-Mason commented:
In terms of the error message, I would agree with @jperson1 that we should drop the "Oops!" for tonal consistency. Also, based on our voice/tone guide, we want our messaging to be authoritative, so I would delete "looks like"

<s>Oops! It looks like some of</s> the information you entered is invalid. Please double-check your information for errors.

From my reading of the ticket, it sounds like this error pops up when a user inputs an invalid End date. If that is the case, can we get more specific about "the information" in the error message, and replace it with something like this: "The date you entered is invalid. Please double-check the **End date** field for errors." 

Open to any suggestions if that sounds too specific!

rnovak338 commented:
> In terms of the error message, I would agree with @jperson1 that we should drop the "Oops!" for tonal consistency. Also, based on our voice/tone guide, we want our messaging to be authoritative, so I would delete "looks like"
> 
> Oops! It looks like some of the information you entered is invalid. Please double-check your information for errors.
> 
> From my reading of the ticket, it sounds like this error pops up when a user inputs an invalid End date. If that is the case, can we get more specific about "the information" in the error message, and replace it with something like this: "The date you entered is invalid. Please double-check the **End date** field for errors."
> 
> Open to any suggestions if that sounds too specific!

Thanks @James-Paul-Mason, I updated the language to be authoritative. Regarding your other point, the changes in this PR will detect form errors beyond just the start/end dates in this form. Because the errors provided by Django are not always clear to the end user, I added some custom context to handle when the start/end dates are invalid. Like so:

![image](https://github.com/user-attachments/assets/1ec50074-7707-4ca3-a968-24594764440d)

I do not think it is possible for the other fields to have invalid input. Either way, moving forward we can provide custom information for them if needed.
